File: apps/salmon-vault-android/src/com/mku/salmon/vault/dialog/SalmonDialog.java
package com.mku.salmon.vault.dialog;

import android.app.Activity;
import android.content.Intent;
import android.text.InputType;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.mku.android.file.AndroidSharedFileObserver;
import com.mku.func.Consumer;
import com.mku.salmon.vault.android.R;
import com.mku.salmon.vault.utils.WindowUtils;
import com.mku.salmonfs.SalmonFile;
import com.mku.utils.SalmonFileUtils;

import com.mku.func.BiConsumer;

import java.util.LinkedHashMap;
import java.util.TreeMap;

/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
public class SalmonDialog {
    public static void promptEdit(String title, String msg, BiConsumer<String, Boolean> OnEdit,
                                  String value, boolean isFileName, boolean readOnly, boolean isPassword,
                                  String option) {
        Activity activity = WindowUtils.getUiActivity();
        WindowUtils.runOnMainThread(() ->
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

            LinearLayout layout = new LinearLayout(activity);
            layout.setPadding(20, 20, 20, 20);
            layout.setOrientation(LinearLayout.VERTICAL);

            TextInputLayout msgText = new TextInputLayout(activity, null,
                    R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
            msgText.setBoxBackgroundMode(TextInputLayout.BOX_BACKGROUND_OUTLINE);
            msgText.setBoxCornerRadii(5, 5, 5, 5);
            msgText.setHint(msg);
            View valueText;
            TextInputEditText typePasswd = null;
            if (isPassword) {
                TextInputLayout typePasswdText = new TextInputLayout(activity, null,
                        R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
                typePasswdText.setPasswordVisibilityToggleEnabled(true);
                typePasswdText.setBoxBackgroundMode(TextInputLayout.BOX_BACKGROUND_OUTLINE);
                typePasswdText.setBoxCornerRadii(5, 5, 5, 5);
                typePasswdText.setHint(msg);

                typePasswd = new TextInputEditText(typePasswdText.getContext());
                typePasswd.setInputType(InputType.TYPE_TEXT_VARIATION_PASSWORD |
                        InputType.TYPE_CLASS_TEXT);
                typePasswd.setText(value);
                typePasswdText.addView(typePasswd);
                valueText = typePasswdText;
            } else {
                TextInputEditText text = new TextInputEditText(msgText.getContext());
                text.setText(value);
                if (readOnly) {
                    text.setInputType(InputType.TYPE_NULL);
                    text.setTextIsSelectable(true);
                } else {
                    text.setInputType(InputType.TYPE_CLASS_TEXT);
                }
                if (isFileName) {
                    String ext = SalmonFileUtils.getExtensionFromFileName(value);
                    if (ext != null && ext.length() > 0)
                        text.setSelection(0, value.length() - ext.length() - 1);
                    else
                        text.setSelection(0, value.length());
                } else {
                    text.selectAll();
                }
                valueText = text;
            }
            LinearLayout.LayoutParams parameters = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);
            msgText.addView(valueText);
            layout.addView(msgText, parameters);
            CheckBox optionCheckBox = new CheckBox(activity);
            if (option != null) {
                optionCheckBox.setText(option);
                layout.addView(optionCheckBox, parameters);
            }
            TextInputEditText finalTypePasswd = typePasswd;
            builder.setPositiveButton(activity.getString(android.R.string.ok), (sender, e) ->
            {
                if (OnEdit != null) {
                    if (isPassword)
                        OnEdit.accept(finalTypePasswd.getText().toString(), optionCheckBox.isChecked());
                    else
                        OnEdit.accept(((EditText) valueText).getText().toString(), optionCheckBox.isChecked());
                }
            });
            builder.setNegativeButton(activity.getString(android.R.string.cancel), (sender, e) ->
            {
                ((AlertDialog) sender).dismiss();
            });

            AlertDialog alertDialog = builder.create();
            alertDialog.setTitle(title);
            alertDialog.setCancelable(true);
            alertDialog.setView(layout);

            if (!activity.isFinishing())
                alertDialog.show();
        });
    }

    public static void promptDialog(String title, String body,
                                    String buttonLabel1, Runnable buttonListener1,
                                    String buttonLabel2, Runnable buttonListener2) {
        if (buttonLabel1 == null)
            buttonLabel1 = "Ok";
        Activity activity = WindowUtils.getUiActivity();
        String finalButtonLabel = buttonLabel1;
        WindowUtils.runOnMainThread(() ->
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
            if (title != null)
                builder.setTitle(title);
            builder.setMessage(body);
            if (finalButtonLabel != null)
                builder.setPositiveButton(finalButtonLabel, (s, e) ->
                {
                    if (buttonListener1 != null)
                        buttonListener1.run();
                });
            if (buttonLabel2 != null)
                builder.setNegativeButton(buttonLabel2, (s, e) ->
                {
                    if (buttonListener2 != null)
                        buttonListener2.run();
                });

            AlertDialog alertDialog = builder.create();
            alertDialog.setTitle(title);
            alertDialog.setCancelable(true);

            if (!activity.isFinishing())
                alertDialog.show();
        });
    }


    public static void promptSingleValue(ArrayAdapter<String> adapter, String title,
                                         int currSelection, BiConsumer<AlertDialog, Integer> onClickListener) {
        Activity activity = WindowUtils.getUiActivity();
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
        if (title != null)
            builder.setTitle(title);
        builder.setSingleChoiceItems(adapter, currSelection, (sender, which) ->
        {
            onClickListener.accept((AlertDialog) sender, which);
        });
        AlertDialog alertDialog = builder.create();
        alertDialog.setTitle(title);
        alertDialog.setCancelable(true);

        if (!activity.isFinishing())
            alertDialog.show();
    }


    public static void promptOpenWith(Intent intent, TreeMap<String, String> apps,
                                      android.net.Uri uri, java.io.File sharedFile,
                                      SalmonFile salmonFile, boolean allowWrite,
                                      Consumer<AndroidSharedFileObserver> OnFileContentsChanged) {
        Activity activity = WindowUtils.getUiActivity();
        String[] names = apps.keySet().toArray(new String[0]);
        String[] packageNames = apps.values().toArray(new String[0]);

        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
        builder.setTitle(activity.getString(R.string.ChooseApp));
        builder.setSingleChoiceItems(names, -1, (sender, which) ->
        {
            try {
                AlertDialog alertDialog = (AlertDialog) sender;
                alertDialog.dismiss();

                int activityFlags = Intent.FLAG_GRANT_READ_URI_PERMISSION;
                if (allowWrite)
                    activityFlags |= Intent.FLAG_GRANT_WRITE_URI_PERMISSION;

                AndroidSharedFileObserver fileObserver = AndroidSharedFileObserver.createFileObserver(sharedFile,
                        salmonFile, OnFileContentsChanged);
                fileObserver.startWatching();

                activity.grantUriPermission(packageNames[which], uri, activityFlags);
                intent.setPackage(packageNames[which]);
                activity.startActivity(intent);
            } catch (Exception ex) {
                Toast.makeText(activity, "Could not start application", Toast.LENGTH_LONG).show();
                ex.printStackTrace();
                sharedFile.delete();
            }
        });
        AlertDialog alert = builder.create();
        alert.show();
    }

    public static void promptDialog(String body) {
        promptDialog("", body, null, null, null, null);
    }

    public static void promptDialog(String title, String body) {
        promptDialog(title, body, null, null, null, null);
    }

    public static void promptDialog(String title, String body, String buttonLabel) {
        promptDialog(title, body, buttonLabel, null, null, null);
    }
}

File: apps/salmon-vault-dotnet-android/src/Dialog/SalmonDialog.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
using Android.Content;
using Android.Widget;
using Google.Android.Material.Dialog;
using Google.Android.Material.TextField;
using Mku.Utils;
using Salmon.Vault.Utils;
using System;
using AndroidX.AppCompat.App;
using Android.Views;
using Activity = Android.App.Activity;
using Mku.Android.File;
using Mku.SalmonFS;
using Salmon.Vault.Extensions;
using System.Collections.Generic;
using System.Linq;
using Salmon.Vault.DotNetAndroid;

namespace Salmon.Vault.Dialog;

public class SalmonDialog
{
    public static void PromptEdit(string title, string msg, Action<string, bool> OnEdit,
        string value = "", bool isFileName = false, bool readOnly = false, bool isPassword = false,
        string option = null)
    {
        Activity activity = WindowUtils.UiActivity;
        WindowUtils.RunOnMainThread(() =>
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

            LinearLayout layout = new LinearLayout(activity);
            layout.SetPadding(20, 20, 20, 20);
            layout.Orientation = Orientation.Vertical;

            TextInputLayout msgText = new TextInputLayout(activity, null,
                Resource.Style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
            msgText.BoxBackgroundMode = TextInputLayout.BoxBackgroundOutline;
            msgText.SetBoxCornerRadii(5, 5, 5, 5);
            msgText.Hint = msg;
            View valueText;
            TextInputEditText typePasswd = null;
            if (isPassword)
            {
                TextInputLayout typePasswdText = new TextInputLayout(activity, null,
            Resource.Style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
                typePasswdText.PasswordVisibilityToggleEnabled = true;
                typePasswdText.BoxBackgroundMode = TextInputLayout.BoxBackgroundOutline;
                typePasswdText.SetBoxCornerRadii(5, 5, 5, 5);
                typePasswdText.Hint = msg;

                typePasswd = new TextInputEditText(typePasswdText.Context);
                typePasswd.InputType = Android.Text.InputTypes.TextVariationPassword |
                              Android.Text.InputTypes.ClassText;
                typePasswd.Text = value;
                typePasswdText.AddView(typePasswd);
                valueText = typePasswdText;
            }
            else
            {
                TextInputEditText text = new TextInputEditText(msgText.Context);
                text.Text = value;
                if (readOnly)
                {
                    text.InputType = Android.Text.InputTypes.Null;
                    text.SetTextIsSelectable(true);
                }
                else
                {
                    text.InputType = Android.Text.InputTypes.ClassText;
                }
                if (isFileName)
                {
                    string ext = SalmonFileUtils.GetExtensionFromFileName(value);
                    if (ext != null && ext.Length > 0)
                        text.SetSelection(0, value.Length - ext.Length - 1);
                    else
                        text.SetSelection(0, value.Length);
                }
                else
                {
                    text.SelectAll();
                }
                valueText = text;
            }
            LinearLayout.LayoutParams parameters = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MatchParent, LinearLayout.LayoutParams.MatchParent);
            msgText.AddView(valueText);
            layout.AddView(msgText, parameters);
            CheckBox optionCheckBox = new CheckBox(activity);
            if (option != null)
            {
                optionCheckBox.Text = option;
                layout.AddView(optionCheckBox, parameters);
            }
            builder.SetPositiveButton(activity.GetString(Android.Resource.String.Ok), (object sender, DialogClickEventArgs e) =>
            {
                if (OnEdit != null)
                {
                    if(isPassword)
                        OnEdit(typePasswd.Text.ToString(), optionCheckBox.Checked);
                    else
                        OnEdit((valueText as EditText).Text.ToString(), optionCheckBox.Checked);
                }
            });
            builder.SetNegativeButton(activity.GetString(Android.Resource.String.Cancel), (object sender, DialogClickEventArgs e) =>
            {
                ((AlertDialog)sender).Dismiss();
            });

            AlertDialog alertDialog = builder.Create();
            alertDialog.SetTitle(title);
            alertDialog.SetCancelable(true);
            alertDialog.SetView(layout);

            if (!activity.IsFinishing)
                alertDialog.Show();
        });
    }

    public static void PromptDialog(string title, string body,
                                    string buttonLabel1 = "Ok", Action buttonListener1 = null,
                                    string buttonLabel2 = null, Action buttonListener2 = null)
    {
        Activity activity = WindowUtils.UiActivity;
        WindowUtils.RunOnMainThread(() =>
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
            if (title != null)
                builder.SetTitle(title);
            builder.SetMessage(body);
            if (buttonLabel1 != null)
                builder.SetPositiveButton(buttonLabel1, (s,e)=>
                {
                    if (buttonListener1 != null)
                        buttonListener1();
                });
        if (buttonLabel2 != null)
            builder.SetNegativeButton(buttonLabel2, (s, e) =>
            {
                if (buttonListener2 != null)
                    buttonListener2();
            });

            AlertDialog alertDialog = builder.Create();
            alertDialog.SetTitle(title);
            alertDialog.SetCancelable(true);

            if (!activity.IsFinishing)
                alertDialog.Show();
        });
    }

    public static void PromptSingleValue(ArrayAdapter<string> adapter, string title,
                                         int currSelection, Action<AlertDialog, int> onClickListener)
    {
        Activity activity = WindowUtils.UiActivity;
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
        if (title != null)
            builder.SetTitle(title);
        builder.SetSingleChoiceItems(adapter, currSelection, (object sender, DialogClickEventArgs e) =>
        {
            onClickListener((AlertDialog)sender, e.Which);
        });
        AlertDialog alertDialog = builder.Create();
        alertDialog.SetTitle(title);
        alertDialog.SetCancelable(true);

        if (!activity.IsFinishing)
            alertDialog.Show();
    }

    public static void PromptOpenWith(Intent intent, SortedDictionary<string, string> apps,
        Android.Net.Uri uri, Java.IO.File sharedFile, SalmonFile salmonFile, bool allowWrite,
        AndroidSharedFileObserver.OnFileContentsChanged OnFileContentsChanged)
    {
        Activity activity = WindowUtils.UiActivity;
        string[] names = apps.Keys.ToArray();
        string[] packageNames = apps.Values.ToArray();

        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
        builder.SetTitle(activity.GetString(Resource.String.ChooseApp));
        builder.SetSingleChoiceItems(names, -1, (object sender, DialogClickEventArgs e) =>
        {
            try
            {
                AlertDialog alertDialog = sender as AlertDialog;
                alertDialog.Dismiss();

                ActivityFlags activityFlags = ActivityFlags.GrantReadUriPermission;
                if (allowWrite)
                    activityFlags |= ActivityFlags.GrantWriteUriPermission;

                AndroidSharedFileObserver fileObserver = AndroidSharedFileObserver.CreateFileObserver(sharedFile,
                    salmonFile, OnFileContentsChanged);
                fileObserver.StartWatching();

                activity.GrantUriPermission(packageNames[e.Which], uri, activityFlags);
                intent.SetPackage(packageNames[e.Which]);
                activity.StartActivity(intent);
            }
            catch (System.Exception ex)
            {
                Toast.MakeText(activity, "Could not start application", ToastLength.Long).Show();
                ex.PrintStackTrace();
                sharedFile.Delete();
            }
        });
        AlertDialog alert = builder.Create();
        alert.Show();
    }

    public static void PromptDialog(string body)
    {
        WindowUtils.RunOnMainThread(() =>
        {
            PromptDialog("", body, "Ok");
        });
    }
}